File: Algorithms/095.UniqueBinarySearchTreesII.java
public class Solution {
    public List<TreeNode> generateTrees(int n) {
        List<TreeNode> result = generateTreesSub(1, n);    
        if (result.size() == 1 && result.get(0) == null) {
            return new ArrayList();
        }
        return result;
    }
    private List<TreeNode> generateTreesSub(int start, int end) {
        List<TreeNode> result = new ArrayList<>();
        if (start > end) {
            result.add(null);
            return result;
        }
        TreeNode root;
        if (start == end) {
            root = new TreeNode(start);
            result.add(root);
            return result;
        }
        for (int i=start; i <= end; i++) {
            List<TreeNode> left = generateTreesSub(start, i-1);
            List<TreeNode> right = generateTreesSub(i+1, end);
            for (int j=0; j< left.size(); j++) {
                for (int k=0; k < right.size(); k++) {
                    root = new TreeNode(i);
                    root.left = left.get(j);
                    root.right = right.get(k);
                    result.add(root);
                }
            }
        }
        return result;
    }
}

File: Algorithms/095.UniqueBinarySearchTreesII.py
class Solution:
    # @param {integer} n
    # @return {TreeNode[]}
    def generateTrees(self, n):
        result = self.generateTreesHelper(1, n)
        if len(result) == 1 and result[0] == None:
            return []
        return result

    def generateTreesHelper(self, start, end):
        result = []
        if start > end:
            result.append(None)
            return result
        if start == end:
            root = TreeNode(start)
            result.append(root)
            return result
        for i in range(start, end+1):
            left = self.generateTreesHelper(start, i-1)
            right = self.generateTreesHelper(i+1, end)
            for j in range(len(left)):
                for k in range(len(right)):
                    root = TreeNode(i)
                    root.left = left[j]
                    root.right = right[k]
                    result.append(root)
        return result
